### PR TITLE
refactor(hooks)!: refactor hooks interface

### PR DESCRIPTION
**Pull request by StarpTech**
https://github.com/wundergraph/wundergraph/blob/feat/refactor_hook_interface/docs/migrations/sdk-0.92.5-0.93.md

#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: pkg/apihandler/internalapihandler.go [hunk 1]
```diff
@@ -211,24 +212,68 @@ func (h *InternalApiHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	}
 
-	ctx.Variables = variablesBuf.Bytes()
+	var request *http.Request
+
+	body := bodyBuf.Bytes()
+
+	// internal requests transmit the client request as a JSON object
+	// this makes it possible to expose the original client request to hooks triggered by internal requests
+	clientRequest, _, _, _ := jsonparser.Get(body, "__wg", "clientRequest")
```
Comment (StarpTech) [inline pkg/apihandler/internalapihandler.go:229]: Move to function

File: pkg/templates/assets/templates/nextjs-starter/.wundergraph/wundergraph.server.ts [hunk 2]
```diff
@@ -3,11 +3,11 @@ import { configureWunderGraphServer } from '@wundergraph/sdk';
 import type { HooksConfig } from './generated/wundergraph.hooks';
 import type { InternalClient } from './generated/wundergraph.internal.client';
 
-export default configureWunderGraphServer<HooksConfig, InternalClient>((serverContext) => ({
+export default configureWunderGraphServer<HooksConfig, InternalClient>(() => ({
 	hooks: {
 		queries: {
 			FakeWeather: {
-				mockResolve: async (hookContext) => {
+				mockResolve: async (hook) => {
```
Comment (jensneuse) [inline pkg/templates/assets/templates/nextjs-starter/.wundergraph/wundergraph.server.ts:10]: Can we add the type of the hook "thing"?

File: pkg/apihandler/internalapihandler.go [hunk 3]
```diff
@@ -211,24 +212,44 @@ func (h *InternalApiHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil && !errors.Is(err, io.EOF) {
+		http.Error(w, "bad request", http.StatusBadRequest)
+		return
+	}
 
-	ctx := pool.GetCtx(r, pool.Config{
-		RenameTypeNames: h.renameTypeNames,
-	})
-	defer pool.PutCtx(ctx)
+	request := r
```
Comment (jensneuse) [inline pkg/apihandler/internalapihandler.go:223]: maybe rename to "internalRequest"?

File: pkg/apihandler/internalapihandler.go [hunk 4]
```diff
@@ -259,3 +280,35 @@ func (h *InternalApiHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		method, err := jsonparser.GetString(body, "__wg", "clientRequest", "method")
+		if err != nil {
+			return nil, err
+		}
+		requestURI, err := jsonparser.GetString(body, "__wg", "clientRequest", "requestURI")
+		if err != nil {
+			return nil, err
+		}
+
+		// create a new request from the client request
```
Comment (jensneuse) [inline pkg/apihandler/internalapihandler.go:296]: add comment why we're excluding the body

File: packages/nextjs-testapp/.wundergraph/wundergraph.server.ts [hunk 5]
```diff
@@ -15,11 +15,31 @@ const testEnum = new GraphQLEnumType({
...
+					user: {
+						name: 'test',
+					},
+				};
+			},
+		},
 		queries: {
 			FakeWeather: {
-				mockResolve: async (hookContext) => {
+				mockResolve: async (hook) => {
```
Comment (jensneuse) [inline packages/nextjs-testapp/.wundergraph/wundergraph.server.ts:42]: add type?

File: packages/default-testapp/.wundergraph/wundergraph.server.ts [hunk 6]
```diff
@@ -16,26 +16,66 @@ import {
 } from 'graphql';
 import { createGraphQLSchema } from 'openapi-to-graphql';
 import jsonPlaceholder from './../json_placeholder.json';
+import { DragonsResponse, SDLResponse } from './generated/models';
 
-export default configureWunderGraphServer<HooksConfig, InternalClient>((serverContext) => ({
+export default configureWunderGraphServer<HooksConfig, InternalClient>(() => ({
 	hooks: {
+		authentication: {
+			postAuthentication: async (hook) => {},
```
Comment (jensneuse) [inline packages/default-testapp/.wundergraph/wundergraph.server.ts:24]: add types for hook arg?

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: pkg/apihandler/internalapihandler.go (hunk 3 above)
Comment (StarpTech) [reply pkg/apihandler/internalapihandler.go:223]: Good point.

File: packages/nextjs-testapp/.wundergraph/wundergraph.server.ts (hunk 5 above)
Comment (StarpTech) [reply packages/nextjs-testapp/.wundergraph/wundergraph.server.ts:42]: This bloats the examples. Every code editor supports inline types and the user has type support anyway.

Comment (StarpTech) [reply packages/nextjs-testapp/.wundergraph/wundergraph.server.ts:42]: I added the types to the official examples but left them in our test apps.

File: docs/migrations/sdk-0.92.5-0.93.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  postAuthentication: async ({ user, internalClient }) => {
+    // let's add a custom hook to update the last login field for the user
+    internalClient.mutations.SetLastLogin({ email: user.email });
+  }
+}
+
+queries: {
+  Dragons: {
+    // A single argument. Here we use object destructuring for better readability.
+    preResolve: async ({ user, log, clientRequest, setClientRequestHeader, internalClient, ...others }) => {};
```
Comment (StarpTech) [inline docs/migrations/sdk-0.92.5-0.93.md:32]: Remove "setClientRequestHeader"

**Pull request merged**